signwriting_animation/diffusion/core/models.py

```diff
@@ -0,0 +1,179 @@
+            num_layers: Number of Transformer layers.
+            num_heads: Number of attention heads.
+            dropout: Dropout probability.
+            ablation: Ablation study parameter.
+            activation: Activation function.
+            legacy: Legacy flag.
+            arch: Architecture type: "trans_enc", "trans_dec", or "gru".
+            cond_mask_prob: Condition mask probability for classifier-free guidance (CFG).
+            device: Device to run the model.
+        """
+        super().__init__()
+
+        self.dims = dims
+        self.keypoints = keypoints
+        self.input_feats = input_feats
+
+        self.ablation = ablation
+        self.cond_mask_prob = cond_mask_prob
+
+        # local conditions
+        self.future_motion_process = MotionProcess(self.input_feats, self.latent_dim)
+        self.past_motion_process = MotionProcess(self.input_feats, self.latent_dim)
+        self.sequence_pos_encoder = PositionalEncoding(self.latent_dim, self.dropout)
+
+        # global conditions
+        self.embed_signwriting = EmbedSignWriting(latent_dim, embedding_arch)
+        self.embed_timestep = TimestepEmbedder(self.latent_dim, self.sequence_pos_encoder)
+
+        if arch == 'trans_enc':
+            print("TRANS_ENC init")
+
+            seqTransEncoderLayer = nn.TransformerEncoderLayer(d_model=latent_dim,
+                                                              nhead=num_heads,
+                                                              dim_feedforward=ff_size,
+                                                              dropout=dropout,
+                                                              activation=activation)
+
+            self.seqEncoder = nn.TransformerEncoder(seqTransEncoderLayer,
+                                                    num_layers=num_layers)
+        elif arch == 'trans_dec':
+            print("TRANS_DEC init")
+            seqTransDecoderLayer = nn.TransformerDecoderLayer(d_model=latent_dim,
+                                                              nhead=num_heads,
+                                                              dim_feedforward=ff_size,
+                                                              dropout=dropout,
+                                                              activation=activation)
+            self.seqEncoder = nn.TransformerDecoder(seqTransDecoderLayer,
+                                                    num_layers=num_layers)
+
+        elif arch == 'gru':
+            print("GRU init")
+            self.seqEncoder = nn.GRU(self.latent_dim, self.latent_dim, num_layers=num_layers, batch_first=True)
+        else:
+            raise ValueError('Please choose correct architecture [trans_enc, trans_dec, gru]')
+
+        self.pose_projection = OutputProcessMLP(self.input_feats, latent_dim, self.keypoints, self.dims)
+
+    def forward(self, x, timesteps, past_motion, signwriting_im_batch):
```

Comment:
could you add types? and possibly shapes in a docstring (similar to the `interface` function, but maybe use gpt to make it a nicer docstring)

Reply:
I've used gpt to generate docstrings. Let me know if the format looks alright. I can change it to a different style if we prefer